core/dbt/adapters/cache.py

```diff
@@ -330,14 +334,15 @@ def add(self, relation):
         :param BaseRelation relation: The underlying relation.
         """
         cached = _CachedRelation(relation)
-        logger.debug('Adding relation: {!s}'.format(cached))
-
-        lazy_log('before adding: {!s}', self.dump_graph)
+        fire_event(AddRelation(relation=cached))
+        # TODO: conditionally fire this event (logger.disabled, if it was std python logger)
+        fire_event(DumpBeforeAddGraph(graph=self.dump_graph()))
```

Comment:
OH I see what `lazy_log` is doing here. I think it's guarding against _log levels_. If we're on info level, but this is debug level, because of strict evaluation, the long running `dump graph` will get called even though the message will never be displayed. The way the event module works right now, we force the message early so lazy logging won't help. I think this might be worth solving inside `fire_event` so we as developers don't have to decide whether the construction of a message is long running or not, it only constructs each message as they are needed.